sqld/src/database/write_proxy.rs

```diff
@@ -118,39 +119,33 @@ impl WriteProxyDatabase {
 
 #[async_trait::async_trait]
 impl Database for WriteProxyDatabase {
-    async fn execute_batch(&self, queries: query::Queries) -> Result<(Vec<QueryResult>, State)> {
+    async fn execute_program(&self, pgm: Program) -> Result<(Vec<Option<QueryResult>>, State)> {
         let mut state = self.state.lock().await;
         if *state == State::Init
-            && queries.iter().all(|q| q.stmt.is_read_only())
-            && final_state(*state, queries.iter().map(|s| &s.stmt)) == State::Init
+            && pgm.is_read_only()
+            && final_state(*state, pgm.steps.iter().map(|s| &s.query.stmt)) == State::Init
```

Comment:
I'm not sure if this is correct: some statements in the program can be skipped, but `final_state()` assumes that we always execute all statements.

Reply:
That's right, I went a bit too fast over this. I pushed a commit that address this issue, lmk